psydac/api/essential_bc.py

```diff
@@ -104,52 +146,126 @@ def apply_essential_bc_2d_StencilMatrix(V, bc, a):
         if e2 == V2.nbasis-1 and bc.boundary.ext == 1:
             a[:,e2-order,:,:] = 0.
 
-
-
 #==============================================================================
-def apply_essential_bc_3d_StencilVector(V, bc, a):
+def apply_essential_bc_3d_StencilMatrix(V, bc, a):
     """ Apply homogeneous dirichlet boundary conditions in 3D """
 
     # assumes a 3D Tensor space
     # add asserts on the space if it is periodic
 
     V1,V2,V3 = V.spaces
 
-    s1, s2, s3 = a.space.starts
-    e1, e2, e3 = a.space.ends
+    s1, s2, s3 = a.domain.starts
+    e1, e2, e3 = a.domain.ends
```

Comment:
Assuming that `a.domain` is a discrete domain, what do `starts` and `ends` mean? And how are they different from the values of `starts` and `ends` in `a.space`?

Reply:
before we had a.domain == a.codomain which mean a.space = a.domain
but if a.domain and a.codomain are different a.space don't have a meaning

Reply:
I see! So our abstract LinearOperator does not have a `space` property anymore?

Reply:
yes

Reply:
Oh yes, there is no `space` anymore: https://github.com/pyccel/psydac/blob/6be6dffd7199ed93a4e83c24244d111ddc5d31bb/psydac/linalg/basic.py#L94

So we had a bug in these two lines of `essential_bc.py` before you changed them. Strange that we missed it.